Support non-covering indexes.

petermattis opened the pull request:
Added indexJoinNode which is a simplified type of join used during index
scans which have to lookup the full row.

Fixes #2190.

petermattis committed:
Support non-covering indexes.

Added indexJoinNode which is a simplified type of join used during index
scans which have to lookup the full row.

Fixes #2190.

vivekmenezes commented on sql/select.go:
```diff
@@ -251,16 +249,18 @@ func (v *indexInfo) init(s *scanNode) {
 	} else {
 		v.cost = 1
+		if !v.covering {
+			v.cost += float64(1 + len(v.desc.Columns) - len(v.desc.PrimaryIndex.ColumnIDs))
```
this can be negative. Are we fine with that?

petermattis replied:
How can it be negative? The primary index cannot have more columns than are in the table.

vivekmenezes replied:
ah sorry read this somehow as subracting number of secondary index columns from the primary index.

vivekmenezes commented on sql/join.go:
```diff
@@ -0,0 +1,146 @@
+// that are looked up in the table.
+type indexJoinNode struct {
+	index            *scanNode
+	table            *scanNode
```
is there value to embedding this so that you don't have to implement Columns(), Values() and Ordering()?

petermattis replied:
Maybe. You're correct that we wouldn't have to implement `{Columns,Values,Ordering}()`, but it doesn't feel conceptually correct. And I'd like `indexJoinNode` to be the template for `joinNode` for which embedding like that almost certainly won't be possible.

vivekmenezes commented on sql/join.go:
```diff
@@ -0,0 +1,146 @@
+	n.table.kvIndex = 0
+	n.table.spans = n.table.spans[0:0]
+
+	for len(n.table.spans) < joinBatchSize {
```
I prefer for i := 0; i < joinBatchSize; i++
just because it's clear how many iterations this is going to take without looking at the append below.

petermattis replied:
Heh, I originally had it that way, but wanted to make it explicit that the loop existed when we had batched up enough spans.

vivekmenezes commented:
LGTM

petermattis committed:
Index joins retain ordering.

Index joins retain the ordering that we scanned the index. Careful setup
of the index scan render targets allow this ordering info to be passed
to sortNode in order to avoid sorting.

petermattis commented on sql/join.go:
```diff
@@ -0,0 +1,172 @@
+			colIDtoRowIndex[col.ID] = len(indexScan.render)
+			indexScan.render = append(indexScan.render, indexScan.getQVal(*col))
+		}
+	}
```
FYI, I added another commit which made the above a bit more complicated. The upshot is that we can avoid sorting when the index scan outputs rows in the desired order.
